### PR TITLE
[12.0][NEW] crm_task module

### PR DESCRIPTION
**Pull request by v-garcies**
I developed this module which links project tasks to CRM opportunities.

This is my first contribution, can you guide me if there are changes to be made?

### CODE REVIEW

**Comment by v-garcies** [comment]
Do I need to do something more?

**Comment by rousseldenis** [comment]
@v-garcies Please check https://runbot3.odoo-community.org/runbot/static/build/3425819-336-cef348/logs/job_20_test_all.txt warnings

**Comment by HaraldPanten** [comment]
Hi @v-garcies are you going to attend @rousseldenis comments?

THX

**Comment by pedrobaeza** [comment]
What is the difference with https://github.com/odoo/odoo/tree/12.0/addons/crm_project ?

**Comment by v-garcies** [comment]
@pedrobaeza This module adds to the CRM opportunities a smart button with the number and direct access to the related tasks

**Comment by pedrobaeza** [comment]
Yeah, but it adds no value IMO to the existing module.

**Comment by v-garcies** [comment]
The goal is to be able to find the tasks related to an opportunity, with the crm_project module this is impossible ... Maybe we should add this improvement to the crm_project module directly ?

**Comment by pedrobaeza** [comment]
If I don't remember bad, in the chatter an entry is created when linked to the task (that should mostly be only one, not several).

**Comment by v-garcies** [comment]
I don't see information about the associated opportunities on the task form. In addition, crm_project acts only on the leads and not on the opportunities.

This module has a different purpose, it allows you to follow the tasks carried out from the opportunity.

Example of opportunity without task :
opportunity-form_without related-task

Example of opportunity with 2 related tasks :
opportunity-form

Example of list of linked tasks by clicking on the smart button :
related-tasks-list

**Comment by pedrobaeza** [comment]
OK, let's see if reviewers find it useful.

**Review by HaraldPanten — APPROVED** [review]
It's functionally working fine. LGTM 👍 

@ValentinVinagre

File: crm_task/models/crm_lead.py
Comment (ValentinVinagre) [inline crm_task/models/crm_lead.py]: Why does it depend on the partner_id if the customer_id isn't used in the '_compute_task_count' function?
Could the taskconunt field  be calculated like this?
```
def _compute_task_count(self):
    self.taskcount = len(self. task_ids)
```

File: crm_task/models/project_task.py
Comment (ValentinVinagre) [inline crm_task/models/project_task.py]: why the required false? by default is false.

**Review by ValentinVinagre — COMMENTED** [review]
Some comments.

File: crm_task/models/crm_lead.py
Comment (v-garcies) [reply crm_task/models/crm_lead.py]: I just made the changes you suggest

File: crm_task/models/project_task.py
Comment (v-garcies) [reply crm_task/models/project_task.py]: I just deleted that, thanks for your feedback

**Comment by v-garcies** [comment]
The Travis CI build failed due to crm_phonecall_planner module :

crm_phonecall_planner/wizards/crm_phonecall_planner.py:171: [E8103(sql-injection), CrmPhonecallPlan.action_accept] SQL injection risk. Use parameters if you can. - More info https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#no-sql-injection

**Comment by pedrobaeza** [comment]
That are not the real errors, but warnings. Yours are:
```
crm_task/models/crm_lead.py:4:1: F401 'odoo.api' imported but unused
crm_task/models/crm_lead.py:18:5: E303 too many blank lines (2)
```

**Comment by v-garcies** [comment]
> That are not the real errors, but warnings. Yours are:
> 
> ```
> crm_task/models/crm_lead.py:4:1: F401 'odoo.api' imported but unused
> crm_task/models/crm_lead.py:18:5: E303 too many blank lines (2)
> ```

Ok sorry, I have a little trouble reading the travis report ...
So, I have corrected the points listed 👍

File: crm_task/models/crm_lead.py
Comment (ValentinVinagre) [inline crm_task/models/crm_lead.py]: the field is required=False by default, it's not necessary.

File: crm_task/views/crm_lead.xml
Comment (ValentinVinagre) [inline crm_task/views/crm_lead.xml]: Why do you show the task twice? Isn't enough with one field?

File: crm_task/views/project_task.xml
Comment (ValentinVinagre) [inline crm_task/views/project_task.xml]: Why do you create a new action? Can't you use one of the existing?

**Review by ValentinVinagre — COMMENTED** [review]
Some comments.

File: crm_task/views/project_task.xml
Comment (v-garcies) [reply crm_task/views/project_task.xml]: I created a new action to remove the automatic filter "My task" and especially to open the tasks in list view by default

If I use one of the existing actions we come across this view by clicking on the task button from an opportunity with no related tasks :
Capture d’écran 2020-07-09 à 12 36 07

File: crm_task/models/crm_lead.py
Comment (v-garcies) [reply crm_task/models/crm_lead.py]: ok, I will remove this attribute 👍

File: crm_task/views/crm_lead.xml
Comment (v-garcies) [reply crm_task/views/crm_lead.xml]: My clients regularly tell me that odoo is great but that the majority of things are too long to perform or require too many clicks.
So I try to improve the user experience by making the information more readable without necessarily having to change views.

However the smart button is still necessary if you want to consult the tasks in calendar view type for example

Do you have a better idea?

Comment (ValentinVinagre) [reply crm_task/views/crm_lead.xml]: The main problem is that with the two fields the information is duplicated, so it should only be shown 1 way, IMO the button is the best option.

File: crm_task/views/project_task.xml
Comment (ValentinVinagre) [reply crm_task/views/project_task.xml]: Okay. A little more, I would move the "opportunity" field down to "Assigned to" in the task view.
Captura de pantalla 2020-07-14 a las 12 33 25

**Comment by HaraldPanten** [comment]
@v-garcies please, squash commits

**Comment by v-garcies** [comment]
> @v-garcies please, squash commits

Hi, ok how should I do this ? 🤔

**Comment by HaraldPanten** [comment]
https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history#reword-other-commit-messages-take-2

git rebase -i HEAD~x ( x: depending on the number of your last commits that you need to squash)

In this example, you have some "administrative" commits, like:

Captura de pantalla 2020-07-21 a las 9 33 37

You should squash them and only keep the commits that make sense like the [ADD] ones, [IMP] ones, important [FIX] ones... 

In you example, keeping commits where you say that you're fixing travis/runbot errors, or where you say that you're taking in consideration ValentinVinagre's comments, doesn't make sense.

Do you know what I mean?

**Comment by v-garcies** [comment]
> https://thoughtbot.com/blog/git-interactive-rebase-squash-amend-rewriting-history#reword-other-commit-messages-take-2
> 
> git rebase -i HEAD~x ( x: depending on the number of your last commits that you need to squash)
> 
> In this example, you have some "administrative" commits, like:
> 
> Captura de pantalla 2020-07-21 a las 9 33 37
> 
> You should squash them and only keep the commits that make sense like the [ADD] ones, [IMP] ones, important [FIX] ones...
> 
> In you example, keeping commits where you say that you're fixing travis/runbot errors, or where you say that you're taking in consideration ValentinVinagre's comments, doesn't make sense.
> 
> Do you know what I mean?

Ok, thanks for the explanations, I'll take care of that

**Comment by HaraldPanten** [comment]
Hi @v-garcies ,

As your module is an [ADD] you should leave only the first commit because you only have 1 relevant commit. Look at this example from another PR:

Captura de pantalla 2020-07-21 a las 11 05 59

So, regarding your commits, you should leave only the first one and you should rename this one following the structure --> [VERSION][ADD] module_name =>(normally) [12.0][ADD] crm_task or something like [12.0][ADD] new module crm_task.
Also a small description of the commit will be welcome; for example: this new module helps users to link tasks to CRM opportunities.

Captura de pantalla 2020-07-21 a las 11 07 03

**Comment by v-garcies** [comment]
@HaraldPanten My first git squash went very badly, I deleted all the last commits even though I used the "fixup" option ... I just corrected all that right now. Thank you for your explanations!

Hope all is well this time :)

**Comment by HaraldPanten** [comment]
Looks better now 👍 

Let's wait for @pedrobaeza , @ValentinVinagre and/or others reviews.

**Comment by pedrobaeza** [comment]
Well, I don't see it needed, but if others do, then those should review.

File: crm_task/readme/USAGE.rst
Comment (ValentinVinagre) [inline crm_task/readme/USAGE.rst]: This feature doesn't exist. You could explain how to link a task with a lead/opportunity.

**Review by ValentinVinagre — CHANGES_REQUESTED** [review]
A little change.

File: crm_task/views/project_task.xml
Comment (ValentinVinagre) [inline crm_task/views/project_task.xml]: As you're working on 'user_id' you don't need to use ref="hr_timesheet.view_task_form2_inherited" because this field is in "project.view_task_form2". Right?

File: crm_task/views/project_task.xml
Comment (ValentinVinagre) [inline crm_task/views/project_task.xml]: Please delete the empty lines

File: crm_task/__manifest__.py
Comment (ValentinVinagre) [inline crm_task/__manifest__.py]: Remove dependency if you change the inherited view.

**Review by ValentinVinagre — COMMENTED** [review]
The last changes :D the last effort.

**Comment by HaraldPanten** [comment]
Hi @v-garcies you should squash commits, as you have 4 different [ADD].

**Review by claudiagn — APPROVED** [review]
functional review ok!

**Comment by pedrobaeza** [comment]
Aborting, as this PR includes a merge commit that messes up the repo. @dreispt please don't merge it so lightly, as others then have to fix the errors this provokes.

**Comment by dreispt** [comment]
Hello @pedrobaeza, merge decision is made based on the PR Reviews, and I see there no strong blocking reason there.
I see no changes that would break the repo, and the build is green. Can you elaborate please?

**Comment by pedrobaeza** [comment]
Merge commits included in a pull request (like the fourth commit here), scrambles the commit history in the repo, having multiple parent ramifications, and when using the `git format-patch` method we used for rescuing commit history when migrating modules, this makes to crash and have conflicts, so always avoid such things. In fact, thanks to our bot that rebases the branch before merging, we are avoiding this problem more frequently.

**Pull request closed without merging**
